adsdata/reader.py

```diff
@@ -0,0 +1,180 @@
+    
+    def process_line(self, line):
+        return line
+    
+    
+   
+class BibcodeFileReader(ADSClassicInputStream):
+    """add id field to bibcode"""
+    
+    def __init__(self, file_):
+        super(BibcodeFileReader, self).__init__(file_)
+        self.counter = 1
+        
+    def process_line(self, line):
+        bibcode = line[:-1]
+        row = '{}\t{}\n'.format(bibcode, self.counter)
+        self.counter += 1
+        return row
+    
+ 
+class RefereedFileReader(ADSClassicInputStream):
+    def __init__(self, file_):
+        super(RefereedFileReader, self).__init__(file_)
+        
+    def process_line(self, line):
+        bibcode = line[:-1]
+        row = '{}\t{}\n'.format(bibcode, 'T')
+        return row
+        
+class StandardFileReader(ADSClassicInputStream):
+    """each line has all values for a single bibcode"""
+    def __init__(self, file_type_, file_):
+        super(StandardFileReader, self).__init__(file_)
+        self.file_type = file_type_
+        # the following lists controls how they are processed
+        # as_array: should values be read in as an array and output to sql as an array
+        #  for example, downloads and grants are an array while relevance has several distinct values but isn't an array
+        self.array_types = ('download', 'reads', 'author', 'reference', 'grants', 'citation', 'reader', 'simbad')
+        # quote_value: should individual values sent to sql be in quotes.  
+        #  for example, we don't quote bibcode, but we do names of authors
+        self.quote_values = ('author','simbad','grants')
+        # tab_separator: is the tab a separator in the input data
+        self.tab_separated_values = ('author')
+        
+    def read(self, size=-1):
+        line = self._iostream.readline()
+        if len(line) == 0:
+            return ''
+        # does the next line match the current bibcode?
+        bibcode = line[:19]
+        value = line[20:-1]
+        match = self._bibcode_match(bibcode)
+        if match and isinstance(value, str):
+            value = [value]
+        while match:
+            line = self._iostream.readline()
+            value.append(line[20:-1])
+            match = self._bibcode_match(bibcode)
+        return self.process_line(bibcode, value)
+    
+
+    def readline(self):
+        line = self._iostream.readline()
+        return self.process_line(line)
+
+        
+    def process_line(self, bibcode, value):
+        as_array = self.file_type in self.array_types
+        quote_value = self.file_type in self.quote_values
+        tab_separator = self.file_type in self.tab_separated_values
+        processed_value = self.process_value(value, as_array, quote_value, tab_separator)
+        row = '{}\t{}\n'.format(bibcode, processed_value)
+        return row
+    
+    def _bibcode_match(self, bibcode):
+        """ peek ahead to next line for bibcode and check for mactch"""
+        file_location = self._iostream.tell()
```

Comment:
Generally each line in any of the data files is pretty short.  I imagine the buffered file reader is mostly moving the pointer to the next byte to read back and forth.  If so, it may be an inexpensive operation.